### PR TITLE
Support more propagation modes with LightStep

### PR DESCRIPTION
**Pull request by rnburn**
Adds additional propagation modes for LightStep's tracer.

Fixes https://github.com/envoyproxy/envoy/issues/10273

### CODE REVIEW

File: api/envoy/config/trace/v2/trace.proto [hunk 1]
```diff
@@ -61,12 +61,30 @@ message Tracing {
...
+    ENVOY = 0;
+
+    // Propagate trace context using LightStep's native format.
+    LIGHTSTEP = 1;
+
+    // Propagate trace context using the b3 format.
+    B3 = 2;
+
+    // Propagation trace context using the w3 trace-context standard.
+    TRACE_CONTEXT = 3;
```
Comment (htuch) [inline api/envoy/config/trace/v2/trace.proto]: Is there a chance that now or in the future any of these formats requires additional parameters? If so, I would favor a `oneof` with a message type per format to future proof.

Comment (rnburn) [reply api/envoy/config/trace/v2/trace.proto]: They shouldn't require more arguments. They correspond to a similar [enum](https://github.com/lightstep/lightstep-tracer-cpp/blob/master/include/lightstep/tracer.h#L25) in the tracer's options.

File: api/envoy/config/trace/v2/trace.proto [hunk 2]
```diff
@@ -61,12 +61,30 @@ message Tracing {
...
+
   // The cluster manager cluster that hosts the LightStep collectors.
   string collector_cluster = 1 [(validate.rules).string = {min_bytes: 1}];
 
   // File containing the access token to the `LightStep
   // <https://lightstep.com/>`_ API.
   string access_token_file = 2 [(validate.rules).string = {min_bytes: 1}];
+
+  // Propagation modes to use by LightStep's tracer.
+  repeated PropagationMode propagation_modes = 3;
```
Comment (mattklein123) [inline api/envoy/config/trace/v2/trace.proto]: I think you want a defined only enum constraint here or above?

**Comment by moderation** [comment]
@rnburn I'm getting immediate crashes on Linux amd64 / aarch64 and MacOS amd64 when trying to launch Envoy with a Lightstep tracer after this PR. The crash occurs whether I include the new `propagation_modes` config stanza or not.

Linux `terminating with uncaught exception of type std::length_error: vector`
MacOS ` libc++abi.dylib: terminating with uncaught exception of type std::length_error: vector`

Removing the Lightstep tracer allows Envoy to launch.
```json
"tracing": {
    "http": {
       "name": "envoy.tracers.lightstep",
       "typed_config": {
          "@type": "type.googleapis.com/envoy.config.trace.v3.LightstepConfig",
          "access_token_file": "/home/<username>/Library/envoy/lightstep_access_token",
          "collector_cluster": "lightstep",
          "propagation_modes": [
            "LIGHTSTEP" ,
            "TRACE_CONTEXT"
          ]
       }
    }
 }
```

**Comment by rnburn** [comment]
I think the problem is this lambda function is invoked outside of the function scope so that propagation_modes becomes a dangling reference.

```
  auto propagation_modes = MakePropagationModes(lightstep_config);

  tls_->set([this, &propagation_modes](
                Event::Dispatcher& dispatcher) -> ThreadLocal::ThreadLocalObjectSharedPtr {
...
}
```

This should be easy to fix. I'll put in a PR.

Here's the backtrace I got from gdb
```
#0  0x00007ffff7c9fed7 in raise () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007ffff7c81535 in abort () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x0000555557781435 in Envoy::TerminateHandler::logOnTerminate() const::{lambda()#1}::operator()() const ()
#3  0x0000555557781443 in Envoy::TerminateHandler::logOnTerminate() const::{lambda()#1}::_FUN() ()
#4  0x000055555814d3f6 in __cxxabiv1::__terminate(void (*)()) ()
#5  0x000055555814d431 in std::terminate() ()
#6  0x000055555814bfc4 in __cxa_throw ()
#7  0x000055555815acb2 in std::__throw_bad_alloc() ()
#8  0x0000555556a5e59c in __gnu_cxx::new_allocator<lightstep::PropagationMode>::allocate(unsigned long, void const*) ()
#9  0x0000555556a5dd46 in std::allocator_traits<std::allocator<lightstep::PropagationMode> >::allocate(std::allocator<lightstep::PropagationMode>&, unsigned long) ()
#10 0x0000555556a5ce92 in std::_Vector_base<lightstep::PropagationMode, std::allocator<lightstep::PropagationMode> >::_M_allocate(unsigned long) ()
#11 0x0000555556a5c7a0 in lightstep::PropagationMode* std::vector<lightstep::PropagationMode, std::allocator<lightstep::PropagationMode> >::_M_allocate_and_copy<__gnu_cxx::__normal_iterator<lightstep::PropagationMode const*, std::vector<lightstep::PropagationMode, std::allocator<lightstep::PropagationMode> > > >(unsigned long, __gnu_cxx::__normal_iterator<lightstep::PropagationMode const*, std::vector<lightstep::PropagationMode, std::allocator<lightstep::PropagationMode> > >, __gnu_cxx::__normal_iterator<lightstep::PropagationMode const*, std::vector<lightstep::PropagationMode, std::allocator<lightstep::PropagationMode> > >) ()
#12 0x0000555556a5baf2 in std::vector<lightstep::PropagationMode, std::allocator<lightstep::PropagationMode> >::operator=(std::vector<lightstep::PropagationMode, std::allocator<lightstep::PropagationMode> > const&) ()
#13 0x0000555556a59b63 in Envoy::Extensions::Tracers::Lightstep::LightStepDriver::LightStepDriver(envoy::config::trace::v3::LightstepConfig const&, Envoy::Upstream::ClusterManager&, Envoy::Stats::Scope&, Envoy::ThreadLocal::SlotAllocator&, Envoy::Runtime::Loader&, std::unique_ptr<lightstep::LightStepTracerOptions, std::default_delete<lightstep::LightStepTracerOptions> >&&, Envoy::Extensions::Tracers::Common::Ot::OpenTracingDriver::PropagationMode, Envoy::Grpc::Context&)::{lambda(Envoy::Event::Dispatcher&)#1}::operator()(Envoy::Event::Dispatcher&) const ()
#14 0x0000555556a5a9a5 in std::_Function_handler<std::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject> (Envoy::Event::Dispatcher&), Envoy::Extensions::Tracers::Lightstep::LightStepDriver::LightStepDriver(envoy::config::trace::v3::LightstepConfig const&, Envoy::Upstream::ClusterManager&, Envoy::Stats::Scope&, Envoy::ThreadLocal::SlotAllocator&, Envoy::Runtime::Loader&, std::unique_ptr<lightstep::LightStepTracerOptions, std::default_delete<lightstep::LightStepTracerOptions> >&&, Envoy::Extensions::Tracers::Common::Ot::OpenTracingDriver::PropagationMode, Envoy::Grpc::Context&)::{lambda(Envoy::Event::Dispatcher&)#1}>::_M_invoke(std::_Any_data const&, Envoy::Event::Dispatcher&) ()
#15 0x0000555556d62cbf in std::function<std::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject> (Envoy::Event::Dispatcher&)>::operator()(Envoy::Event::Dispatcher&) const ()
#16 0x0000555556d5c41b in Envoy::ThreadLocal::InstanceImpl::Bookkeeper::set(std::function<std::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject> (Envoy::Event::Dispatcher&)>)::{lambda(Envoy::Event::Dispatcher&)#1}::operator()(Envoy::Event::Dispatcher&) const ()
#17 0x0000555556d5f742 in std::_Function_handler<std::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject> (Envoy::Event::Dispatcher&), Envoy::ThreadLocal::InstanceImpl::Bookkeeper::set(std::function<std::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject> (Envoy::Event::Dispatcher&)>)::{lambda(Envoy::Event::Dispatcher&)#1}>::_M_invoke(std::_Any_data const&, Envoy::Event::Dispatcher&) ()
#18 0x0000555556d62cbf in std::function<std::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject> (Envoy::Event::Dispatcher&)>::operator()(Envoy::--Type <RET> for --T----Type <RET> for more, q to quit, c to continue without paging--
Event::Dispatcher&) const ()
#19 0x0000555556d5dcac in Envoy::ThreadLocal::InstanceImpl::SlotImpl::set(std::function<std::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject> (Envoy::Event::Dispatcher&)>)::{lambda()#1}::operator()() const ()
#20 0x0000555556d5fd04 in std::_Function_handler<void (), Envoy::ThreadLocal::InstanceImpl::SlotImpl::set(std::function<std::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject> (Envoy::Event::Dispatcher&)>)::{lambda()#1}>::_M_invoke(std::_Any_data const&) ()
#21 0x00005555559feb78 in std::function<void ()>::operator()() const ()
#22 0x0000555556e91df3 in Envoy::Event::DispatcherImpl::runPostCallbacks() ()
#23 0x0000555556e91cc2 in Envoy::Event::DispatcherImpl::run(Envoy::Event::Dispatcher::RunType) ()
#24 0x0000555556e79072 in Envoy::Server::WorkerImpl::threadRoutine(Envoy::Server::GuardDog&) ()
#25 0x0000555556e78916 in Envoy::Server::WorkerImpl::start(Envoy::Server::GuardDog&)::{lambda()#1}::operator()() const ()
#26 0x0000555556e79916 in std::_Function_handler<void (), Envoy::Server::WorkerImpl::start(Envoy::Server::GuardDog&)::{lambda()#1}>::_M_invoke(std::_Any_data const&) ()
#27 0x00005555559feb78 in std::function<void ()>::operator()() const ()
#28 0x00005555577d89a2 in Envoy::Thread::ThreadImplPosix::ThreadImplPosix(std::function<void ()>)::{lambda(void*)#1}::operator()(void*) const ()
#29 0x00005555577d89c6 in Envoy::Thread::ThreadImplPosix::ThreadImplPosix(std::function<void ()>)::{lambda(void*)#1}::_FUN(void*) ()
#30 0x00007ffff7e50182 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#31 0x00007ffff7d79b1f in clone () from /lib/x86_64-linux-gnu/libc.so.6
```

**Comment by rnburn** [comment]
I put in #10562 with a fix

**Pull request merged**